concepts/type-theory-glossary.md

```diff
@@ -0,0 +1,396 @@
+and make sure only a known number of types in the same package implement it.
+(See https://github.com/BurntSushi/go-sumtype for tooling to help support this;
+the readme also contains good documentation of the concept!)
+
+In Rust, the `enum` keyword (confusingly) also creates sum types.
+Rust's sum types are at least properly closed (unlike Java's and Go's).
+
+#### maybe types
+
+In *many* languages, you may find a type in the standard library called "Maybe",
+or equally frequently, "Optional", or, "Either".
+These are a sum type!
+They're typically explained as "either **none**, or **some** of a thing":
+so, they take the cardinality of a thing and simply "plus one" to it,
+and colloquially call that additional state the "none" state.
+
+For some `Maybe<T>`, the cardinality is "1+|T|".
+
+(This is also a great example for how we can use cardinality counting methods
+to create clear comparisons between differently named but functionally
+identical concepts between programming language communities!)
+
+### product types in the wild
+
+In Java, `class` creates a product type.
+
+In Go, `struct` creates a product type.
+
+In Rust, `struct` creates a product type.
+
+### interfaces in the wild
+
+In Java, the `interface` keyword introduces an interface.
+These interfaces are [*explicit*](#explicit-versus-structural-interfaces) -- a class must use the `implements` keyword to explicitly state that it can satisfy that interface.
+(That the class has sufficient methods to match the interface is then also checked at compile time -- but having the methods is not necessary rather than sufficient for the "implements" relationship to be considered true).
+
+In Go, the `interface` keyword introduces an interface.
+These interfaces are [*structural*](#explicit-versus-structural-interfaces) -- a struct doesn't need to *say* it implements an interface;
+if it has the relevant methods, then you can start using values of that type wherever the interface is required.
+
+In Rust, the `trait` feature combined with the `impl`+`for` keywords is most comparable to the concept of interface in other languages.
+Implementing a trait is [*explicit*](#explicit-versus-structural-interfaces) -- a type must use the `impl`+`for` keywords to explicitly state the methods that satisfy the trait.
+
+### what about numbers?
+
+Numbers are... interesting.
+
+For the most part in computing, we tend to pretend and build abstractions as if
+numbers are [infinite scalars](#infinite-scalars).
+
+Of course, this is rarely true in practice.
+Integers are often implemented in such a way that values that would require beyond 32 (or 64, etc) bits are unrepresentable;
+arguably, this makes them more like a (very (*very*) large) enum than anything else.
+Floating point numbers implemented in IEE764 style also have their own entire host of foibles.
+
+The scope of this document is not to solve this conundrum, nor to try to convince you how you should consider numbers.
```

Comment:
As much as I enjoy  reading elaborate English and learning new words, I think for this docs we should keep it simple :) I've heard "conundrum" for the first time, rephrasing it would be appreciate.